[CI Environment] Added scheduled pipelines for Subscription & Management-Group Deployment removals (#2310)
Azure/ResourceModules: docs/wiki/The CI environment - Deployment removal.md

> This pipeline and its scripts can be used to limit the number of deployments that are listed in Azure at any given time. This is especially useful for a scope like 'Management Group' is Azure does not provide any [automatic cleanup](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/deployment-history-deletions?tabs=azure-powershell).

Reviewer: Should this be added to any TOC? maybe the main CI environment page?

Author: Added

The corresponding error message would look similar to `Creating the deployment '<deploymentName>' would exceed the quota of '800'. The current deployment count is '804'. Please delete some deployments before creating a new one, or see https://aka.ms/800LimitFix for information on managing deployment limits.`

The platform pipeline `platform.deployment.removal.yml` has an automatic schedule that runs every night, but can also be triggered on-demand.

> By default, the pipeline will cleanup both the Subscription, as well as Management Group scope using the used script's default configuration. In addition, you can specify a pipeline variable `maxDeploymentRetentionInDays` to define beyond which time frame deployments should always be deleted. Within the timeframe, only non-failed and non-running deployments are considered.

Reviewer: As mentioned, I'd suggest clarifying permissions needed by the SP to run the pipeline as set by default, and mention how to set that up to only clean up subscription scope deployments

Author: Updated

---

### _Navigation_

- [Location](#location)
- [How it works](#how-it-works)
- [How to use it](#how-to-use-it)

---
# Location

You can find the used scripts under
- [`/utilities/pipelines/deploymentRemoval/Clear-SubscriptionDeployment`](https://github.com/Azure/ResourceModules/blob/main/utilities/pipelines/deploymentRemoval/Clear-SubscriptionDeployment.ps1) and
- [`/utilities/pipelines/deploymentRemoval/Clear-ManagementGroupDeployment`](https://github.com/Azure/ResourceModules/blob/main/utilities/pipelines/deploymentRemoval/Clear-ManagementGroupDeployment.ps1)

# How it works

Both scripts work in the same way:

> 1. The script fetches all current deployments from Azure.

Reviewer: Not all, but the ones at MG and sub scope

Author: Updated

> 1. By default it then filters them down to non-running & non-failing deployments (can be modified) unless they are older than a provided time limit in which case they'd always be included.

Reviewer: Are successful deployments within the max retention removed or kept?

Author: Removed. Should now be clearer with the 'number' example

> 1. Lastly, it removes all matching deployments in chunks of 100 deployments each.

Reviewer: I'd suggest an example with numbers

Author: Added

> This pipeline and its scripts can be used to limit the number of deployments that are listed in Azure at any given time. This is especially useful for a scope like 'Management Group' is Azure does not provide any [automatic cleanup](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/deployment-history-deletions?tabs=azure-powershell).

Reviewer: ```suggestion
This pipeline and its scripts can be used to limit the number of deployments that are listed in the Azure deployments history at any given time. This is especially useful for a scope like 'Management Group', since for that scope, Azure Resource Manager does not currently provide [automatic deletion](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/deployment-history-deletions?tabs=azure-powershell) of deployments from your history as you near the 800 limit.
```

> The corresponding error message would look similar to `Creating the deployment '<deploymentName>' would exceed the quota of '800'. The current deployment count is '804'. Please delete some deployments before creating a new one, or see https://aka.ms/800LimitFix for information on managing deployment limits.`

Reviewer: ```suggestion
Without automatic deletion, if the scope you are deploying to is already at the 800 limit, your next deployment fails with an error. The corresponding error message would look similar to `Creating the deployment '<deploymentName>' would exceed the quota of '800'. The current deployment count is '804'. Please delete some deployments before creating a new one, or see https://aka.ms/800LimitFix for information on managing deployment limits.`
```

> - Not register the `platform.deployment.removal.yml` pipeline OR
> - Adjust the pipeline's default behavior to not remove deployments on the scope you'd like to preserve (i.e., change `default: true` to `default: false`)

---

### _Navigation_

- [Location](#location)
- [How it works](#how-it-works)
- [How to use it](#how-to-use-it)

---
# Location

You can find the used scripts under
- [`/utilities/pipelines/deploymentRemoval/Clear-SubscriptionDeployment`](https://github.com/Azure/ResourceModules/blob/main/utilities/pipelines/deploymentRemoval/Clear-SubscriptionDeployment.ps1) and
- [`/utilities/pipelines/deploymentRemoval/Clear-ManagementGroupDeployment`](https://github.com/Azure/ResourceModules/blob/main/utilities/pipelines/deploymentRemoval/Clear-ManagementGroupDeployment.ps1)

# How it works

Both scripts work in the same way:

1. The script fetches all current deployments at a given scope from Azure (i.e., Management-Group-scope or Subscription-scope).
   - For example, it may find
     - 120 successful deployments
     - 10 failed deployments that are 3 weeks old
     - 10 failed deployments that are 1 week old
     - 10 running

> 1. By default it then filters them down to non-running & non-failing deployments (can be modified). It only considers failed deployments if they are older than a provided time limit in which case they are included too.

Reviewer: ```suggestion
1. By default, it then filters them down to non-running & non-failing deployments (can be modified). It only considers failed deployments for deletion if they are older than a provided time limit.
```

>    - Following the example, and with a time limit of 2 weeks, it now only considers the

Reviewer: ```suggestion
   - Following the example, and with a `maxDeploymentRetentionInDays` of 14 days (2 weeks), the following deployments will be targeted for deletion from the deployment history:
```

     -  120 successful deployments
     -  10 failed deployments that are 3 weeks old
1. Lastly, it removes all matching deployments in chunks of 100 deployments each.

>    - In context of the example, this mean it deletes the deployments with to commands, first removing 100, and then 30 deployments.

Reviewer: ```suggestion
   - In the context of the example, this means the procedure deletes the deployments in 2 chunks, first removing 100, and then 30 deployments.
```

> # How to use it

Reviewer: Should this refer to the pipeline usage instead of referring to the scripts?

Author: Yes and no - I was wondering the same things. These sections usually referred to the scripts - not the pipelines. So I tried a 'hybrid', I guess :D

> This pipeline and its scripts can be used to limit the number of deployments that are listed in the Azure deployments history at any given time. This is especially useful for a scope like 'Management Group', since for that scope, Azure Resource Manager does not currently provide [automatic deletion](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/deployment-history-deletions?tabs=azure-powershell) of deployments from your history as you near the 800 limit.

Reviewer: Don't forget to rename the page to `The CI environment - Deployment history cleanup` to keep the link reference
